### PR TITLE
Add cluster definition for smoke-co-alarm

### PR DESCRIPTION
**Pull request by ericzijian1994**
This PR adds an implementation of a Smoke CO Alarm.

### CODE REVIEW

File: examples/smoke-co-alarm-app/silabs/efr32/.gn
Comment (bzbarsky-apple) [inline examples/smoke-co-alarm-app/silabs/efr32/.gn:1]: Please split out the XML additions and core regeneration into a separate PR.  They need to be _very_ carefully reviewed, and are too easy to lose in all the other changes here.

File: src/app/zap-templates/zcl/data-model/chip/concentration-measurement-cluster.xml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+limitations under the License.
+-->
+<configurator>
+  <domain name="CHIP" />
+  <cluster>
+    <name>Concentration Measurement</name>
+    <domain>Measurement &amp; Sensing</domain>
+    <description>Attributes and commands for configuring the measurement of concentration, and
+      reporting concentration measurements.</description>
+    <code>0x040C</code>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/concentration-measurement-cluster.xml:24]: This is the Carbon Monoxide Concentration Measurement cluster (not clearly labeled so in the spec, sadly), not the generic Concentration Measurement.

File: src/app/zap-templates/zcl/data-model/chip/concentration-measurement-cluster.xml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    <define>CONCENTRATION_MEASUREMENT_CLUSTER</define>
+    <client tick="false" init="false">true</client>
+    <server tick="false" init="false">true</server>
+    <globalAttribute side="either" code="0xFFFD" value="1" />
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/concentration-measurement-cluster.xml:28]: Spec says 3, no?  Going to stop here; please break out the XML changes into a separate PR so we can get the XML aligned to the spec without noise from all the other bits.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Comment by bzbarsky-apple** [comment]
@ericzijian1994 First, I can handle the availability.yaml bits, so no need to touch them.

Second, why are you closing this?  That's going to make the review comment tracking very hard.

File: src/app/zap-templates/zcl/data-model/chip/concentration-measurement-cluster.xml (hunk 2 above)
Comment (ericzijian1994) [reply src/app/zap-templates/zcl/data-model/chip/concentration-measurement-cluster.xml:28]: Clear, will open a new PR for XML changes.

**Comment by bzbarsky-apple** [comment]
Ah, I guess you'll do separate PRs for the XML and the other bits, right.  In that case, closing probably makes sense if you mention this one in the others.

**Comment by ericzijian1994** [comment]
@bzbarsky-apple, I have updated the content.

The major new folders are:

1. examples/smoke-co-alarm-app
2. src/app/clusters/smoke-co-alarm-server

In addition, I updated the smoke-co-alarm.xml. #26705

**Comment by bzbarsky-apple** [comment]
> In addition, I updated the smoke-co-alarm.xml. https://github.com/project-chip/connectedhomeip/issues/26705

At risk of sounding like a broken record: please put the XML changes in a separate pr @ericzijian1994

**Comment by ericzijian1994** [comment]
@bzbarsky-apple, I have recommitted the code.

**Comment by ericzijian1994** [comment]
@bzbarsky-apple, now using the new XML that was merged into the master.

File: examples/smoke-co-alarm-app/silabs/efr32/src/ZclCallbacks.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+void MatterPostAttributeChangeCallback(const chip::app::ConcreteAttributePath & attributePath, uint8_t type, uint16_t size,
+                                       uint8_t * value)
+{
+    ClusterId clusterId     = attributePath.mClusterId;
+    AttributeId attributeId = attributePath.mAttributeId;
+    ChipLogProgress(Zcl, "Cluster callback: " ChipLogFormatMEI, ChipLogValueMEI(clusterId));
+
+    if (clusterId == SmokeCoAlarm::Id && attributeId == SmokeCoAlarm::Attributes::TestInProgress::Id)
+    {
+        AlarmMgr().StartSelfTesting();
```
Comment (bzbarsky-apple) [inline examples/smoke-co-alarm-app/silabs/efr32/src/ZclCallbacks.cpp:43]: What if TestInProgress is being set to false?

File: examples/smoke-co-alarm-app/silabs/efr32/src/ZclCallbacks.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    else if (clusterId == Identify::Id)
+    {
+        ChipLogProgress(Zcl, "Identify attribute ID: " ChipLogFormatMEI " Type: %u Value: %u, length %u",
+                        ChipLogValueMEI(attributeId), type, *value, size);
```
Comment (bzbarsky-apple) [inline examples/smoke-co-alarm-app/silabs/efr32/src/ZclCallbacks.cpp:48]: This log line only makes sense if size is known to be 1.

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+        // If the value is busy then do nothing
+        if (expressedState == ExpressedStateEnum::kSmokeAlarm || expressedState == ExpressedStateEnum::kCOAlarm ||
+            expressedState == ExpressedStateEnum::kTesting || expressedState == ExpressedStateEnum::kInterconnectSmoke ||
+            expressedState == ExpressedStateEnum::kInterconnectCO)
+        {
+            success = false;
+        }
+        else
+        {
+            success = SetTestInProgress(commandPath.mEndpointId, true);
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:294]: Spec says ExpressedState should be set to Testing when test starts and then cleared when test stops.  Is the application responsible for doing that?  If so, that should be documented in the header API documentation.

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+template <typename T>
+bool SmokeCoAlarmServer::GetAttribute(EndpointId endpointId, AttributeId attributeId,
+                                      EmberAfStatus (*getFn)(EndpointId endpointId, T * value), T & value) const
+{
+    EmberAfStatus status = getFn(endpointId, &value);
+    bool success         = (EMBER_ZCL_STATUS_SUCCESS == status);
+
+    if (!success)
+    {
+        ChipLogError(Zcl, "Failed to read SmokeCOAlarm attribute: attribute=0x%" PRIx32 ", status=0x%x", attributeId,
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:323]:
```suggestion
        ChipLogError(Zcl, "Failed to read SmokeCOAlarm attribute: attribute=" ChipLogFormatMEI ", status=0x%x", ChipLogValueMEI(attributeId),
```

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+template <typename T>
+bool SmokeCoAlarmServer::SetAttribute(EndpointId endpointId, AttributeId attributeId,
+                                      EmberAfStatus (*setFn)(EndpointId endpointId, T value), T value)
+{
+    EmberAfStatus status = setFn(endpointId, value);
+    bool success         = (EMBER_ZCL_STATUS_SUCCESS == status);
+
+    if (!success)
+    {
+        ChipLogError(Zcl, "Failed to write SmokeCOAlarm attribute: attribute=0x%" PRIx32 ", status=0x%x", attributeId,
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:338]:
```suggestion
        ChipLogError(Zcl, "Failed to write SmokeCOAlarm attribute: attribute=" ChipLogFormatMEI ", status=0x%x", ChipLogValueMEI(attributeId),
```

**Comment by ericzijian1994** [comment]
@bzbarsky-apple I've modified them, added documentation to the headers, and submitted a spec issue.

**Comment by ericzijian1994** [comment]
@bzbarsky-apple I have revised the conflicting content, please review, thanks.

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandHandler.h>
+#include <app/ConcreteCommandPath.h>
+#include <app/util/af.h>
+#include <app/util/config.h>
+#include <protocols/interaction_model/StatusCode.h>
+
+#ifndef SMOKE_CO_ALARM_SERVER_ENDPOINT
+#define SMOKE_CO_ALARM_SERVER_ENDPOINT 1
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h:34]: What is this define and why is it here?  It seems completely unused.

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    typedef bool (*RemoteOpHandler)(chip::EndpointId endpointId);
+
+    /**
+     * @brief Updates the attribute with new value
+     *
+     * @note The application is responsible for changing the expression state
+     *
+     * @param endpointId ID of the endpoint
+     * @param newExpressedState new expressed state
+     * @return true on success, false on failure
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h:62]: What I meant is that one comment like "For all the functions below, this is what the return value means" would be OK.  This comment just documents a single function...

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+
+// =============================================================================
+// Cluster commands callbacks
+// =============================================================================
+
+bool emberAfSmokeCoAlarmClusterSelfTestRequestCallback(CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
+                                                       const Commands::SelfTestRequest::DecodableType & commandData)
+{
+    return SmokeCoAlarmServer::Instance().HandleRemoteSelfTestRequest(commandObj, commandPath,
+                                                                      emberAfPluginSmokeCoAlarmSelfTestRequestCommand);
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:457]: If we are hardcoding this anyway, why not have HandleRemoteSelfTestRequest call this function directly instead of passing the pointer, etc?  Either way, though.

File: examples/smoke-co-alarm-app/silabs/src/ZclCallbacks.cpp [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+void MatterPostAttributeChangeCallback(const chip::app::ConcreteAttributePath & attributePath, uint8_t type, uint16_t size,
+                                       uint8_t * value)
+{
+    ClusterId clusterId     = attributePath.mClusterId;
+    AttributeId attributeId = attributePath.mAttributeId;
+    ChipLogProgress(Zcl, "Cluster callback: " ChipLogFormatMEI, ChipLogValueMEI(clusterId));
+
+    if (clusterId == SmokeCoAlarm::Id && attributeId == SmokeCoAlarm::Attributes::ExpressedState::Id)
+    {
+        SmokeCoAlarm::ExpressedStateEnum expressedState = *(reinterpret_cast<SmokeCoAlarm::ExpressedStateEnum *>(value));
```
Comment (bzbarsky-apple) [inline examples/smoke-co-alarm-app/silabs/src/ZclCallbacks.cpp:43]: Please `static_assert(sizeof(SmokeCoAlarm::ExpressedStateEnum) == 1, "Wrong size");` here, since you are assuming that.

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+        // If the value is busy then return busy
+        if (expressedState == ExpressedStateEnum::kSmokeAlarm || expressedState == ExpressedStateEnum::kCOAlarm ||
+            expressedState == ExpressedStateEnum::kTesting || expressedState == ExpressedStateEnum::kInterconnectSmoke ||
+            expressedState == ExpressedStateEnum::kInterconnectCO)
+        {
+            status = Status::Busy;
+        }
+        else
+        {
+            success = opHandler(endpointId);
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:390]: Again: what is responsible for setting ExpressedState to "Testing"?  If it's the application then:

1. The example app being added here should actually do it!
2. That needs to be very clearly documented on the callback declaration.

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+
+    return success;
+}
+
+bool SmokeCoAlarmServer::SetEndOfServiceAlert(EndpointId endpointId, EndOfServiceEnum newEndOfServiceAlert)
+{
+    bool success =
+        SetAttribute(endpointId, Attributes::EndOfServiceAlert::Id, Attributes::EndOfServiceAlert::Set, newEndOfServiceAlert);
+
+    if (success && (newEndOfServiceAlert == EndOfServiceEnum::kExpired))
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp]: Spec issue is https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/7091

Next time, @ericzijian1994 please mention the link when filing a spec issue?

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp (hunk 12 above)
Comment (ericzijian1994) [reply src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp]: I have now changed to set ExpressedState at the server, so I have added the ExpressedState priority implementation.

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h [hunk 14]
```diff
@@ -158,6 +168,11 @@ class SmokeCoAlarmServer
         chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
         const chip::app::Clusters::SmokeCoAlarm::Commands::SelfTestRequest::DecodableType & commandData);
 
+    /**
+     * @brief Expression status record values
+     */
+    int expressedStateMask = 1;
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h:174]: Why is this signed?  and why is it not named with "m...."?

And the description of this member does not make sense...

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h [hunk 15]
```diff
@@ -102,6 +104,15 @@ class SmokeCoAlarmServer
 
     inline bool SupportsCOAlarm(chip::EndpointId endpointId) { return GetFeatures(endpointId).Has(Feature::kCoAlarm); }
 
+    /**
+     * @brief ExpressedState priority, the priority order of conditions is determined by the manufacturer
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.h:108]: If it's determined by the manufacturer, why is it hardcoded here?  This does not make sense.

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp [hunk 16]
```diff
@@ -55,8 +55,21 @@ SmokeCoAlarmServer & SmokeCoAlarmServer::Instance()
...
 bool SmokeCoAlarmServer::SetExpressedState(EndpointId endpointId, ExpressedStateEnum newExpressedState)
 {
+    expressedStateMask |= (1 << to_underlying(newExpressedState));
+
     ExpressedStateEnum expressedState;
     bool success = GetAttribute(endpointId, Attributes::ExpressedState::Id, Attributes::ExpressedState::Get, expressedState);
+    expressedStateMask |= (1 << to_underlying(expressedState));
+
+    // Get the highest priority alarm
+    for (ExpressedStateEnum i : expressedStatePriority)
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:65]: This logic needs to be in the app, not in the cluster.  Why are these changes being made?

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp [hunk 17]
```diff
@@ -387,14 +443,23 @@ bool SmokeCoAlarmServer::HandleRemoteSelfTestRequest(CommandHandler * commandObj
         }
         else
         {
-            success = opHandler(endpointId);
-            if (!success)
+            success = SetExpressedState(endpointId, ExpressedStateEnum::kTesting);
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:446]: Is this why the priority stuff was added? 

My understanding of the spec is that expressed state unconditionally changes to kTesting when this command is successful, which is not what the priority stuff implements, right?

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp [hunk 18]
```diff
@@ -387,14 +443,23 @@ bool SmokeCoAlarmServer::HandleRemoteSelfTestRequest(CommandHandler * commandObj
...
             {
-                status = Status::Failure;
+                success = SetTestInProgress(endpointId, true);
+            }
+
+            if (success)
+            {
+                success = emberAfPluginSmokeCoAlarmSelfTestRequestCommand(endpointId);
+
+                UnsetExpressedState(endpointId, ExpressedStateEnum::kTesting);
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:456]: This is wrong.  The expressed state needs to change back _when the test is done_.  Which means the app needs to do it.

File: src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp [hunk 19]
```diff
@@ -387,14 +443,23 @@ bool SmokeCoAlarmServer::HandleRemoteSelfTestRequest(CommandHandler * commandObj
...
+                SetTestInProgress(endpointId, false);
```
Comment (bzbarsky-apple) [inline src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp:457]: Again, the app needs to do that.  We don't know when the test is done.

If we want to make this easier, we could have a function for "test is done, here is the new expressed state" that the app would call, but all that would do is SetExpressedState and SetTestInProgress(false).

Comment (ericzijian1994) [reply src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp]: OK, change to APP to set the state.

File: examples/smoke-co-alarm-app/silabs/src/SmokeCoAlarmManager.cpp [hunk 20]
```diff
@@ -37,10 +37,29 @@ CHIP_ERROR SmokeCoAlarmManager::Init()
...
     bool success = true;
+    ExpressedStateEnum expressedState;
 
-    SILABS_LOG("Start self-testing!");
+    success = SmokeCoAlarmServer::Instance().GetExpressedState(1, expressedState);
 
-    SILABS_LOG("End self-testing!");
+    if (success)
+    {
+        success = SmokeCoAlarmServer::Instance().SetExpressedState(1, ExpressedStateEnum::kTesting);
```
Comment (bzbarsky-apple) [inline examples/smoke-co-alarm-app/silabs/src/SmokeCoAlarmManager.cpp:46]: Again: _this_ part should live in the cluster, because all applications will need to do it.

The app knows what its expressed state is (the attr store is not the source of truth for that), and once the self-test is done should update the expressed state to the right value.  But setting it to kTesting should happen in the cluster itself, I would think.

Comment (ericzijian1994) [reply examples/smoke-co-alarm-app/silabs/src/SmokeCoAlarmManager.cpp]: I have moved this part to cluster.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
